File: backend/app/routers/admin_diagnostic.py
"""Admin router for diagnostic item management (Task 3).

Mounted under /admin by admin.py (which already enforces get_current_admin).
The generate and verify-sweep endpoints are additionally rate-limited (LLM calls).

Lifecycle:
    draft  ─► approved  (approve)
    draft  ─► retired   (reject — kept for audit, never served)
    approved ─► retired (retire — e.g. non-discriminating item)
    approved ─► draft   (unpublish — to fix & re-approve)

Plus an in-place flag action that does NOT change status:
    approved (flagged) ─► approved (clear-verifier-flag — dismiss a false positive)
"""
from __future__ import annotations

import logging
import uuid

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_session
from app.core.rate_limit import limiter
from app.models.user import User
from app.routers.admin_auth import get_current_admin
from app.schemas.diagnostic import (
    DiagnosticGenerateRequest,
    DiagnosticItemPatch,
    DiagnosticItemRead,
    DiagnosticListResponse,
    DiagnosticSweepRequest,
    DiagnosticSweepResponse,
    FlaggedItem,
)
from app.services.diagnostic_item_service import (
    approve_item,
    clear_verifier_flag,
    generate_items,
    get_item,
    list_items,
    patch_item,
    reject_item,
    retire_item,
    run_verify_sweep,
    unpublish_item,
)

logger = logging.getLogger(__name__)

router = APIRouter()

# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/verify  (sweep)
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/verify", response_model=DiagnosticSweepResponse)
@limiter.limit("5/minute")
async def verify_sweep_diagnostic_items(
    request: Request,
    payload: DiagnosticSweepRequest,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticSweepResponse:
    """Run the independent answer-verifier over matching diagnostic items.

    Selects items using the supplied filters (market_code, topic, status,
    only_unverified) bounded by ``limit``, runs ``verify_item`` on each
    (best-effort — one LLM failure sets verifier_status="error" and never
    aborts the sweep), then commits and returns counts + the list of flagged
    (mismatch + ambiguous) items.

    This endpoint NEVER changes answer_index or status — advisory only.
    """
    result = await run_verify_sweep(
        session,
        status=payload.status,
        market_code=payload.market_code,
        topic=payload.topic,
        limit=payload.limit,
        only_unverified=payload.only_unverified,
        tier=payload.tier,
    )
    await session.commit()
    return DiagnosticSweepResponse(
        verified=result["verified"],
        agree=result["agree"],
        mismatch=result["mismatch"],
        ambiguous=result["ambiguous"],
        error=result["error"],
        flagged=[FlaggedItem(**f) for f in result["flagged"]],
    )


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/generate
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/generate", response_model=list[DiagnosticItemRead])
@limiter.limit("20/minute")
async def generate_diagnostic_items(
    request: Request,
    payload: DiagnosticGenerateRequest,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> list[DiagnosticItemRead]:
    """Generate draft diagnostic items via the LLM.

    Admin-gated. The limit is 20/min (not 5) so an operator can bulk-generate
    several topics for a market/tier in one go — the admin UI fires one request
    per selected topic (kept per-topic so each stays well under the gateway timeout).
    """
    items = await generate_items(
        session,
        market_code=payload.market_code,
        topic=payload.topic,
        difficulty_tier=payload.difficulty_tier,
        count=payload.count,
    )
    await session.commit()
    return [DiagnosticItemRead.model_validate(i) for i in items]


# ---------------------------------------------------------------------------
# GET /admin/diagnostic-items
# ---------------------------------------------------------------------------


@router.get("/diagnostic-items", response_model=DiagnosticListResponse)
async def list_diagnostic_items(
    market_code: str | None = None,
    topic: str | None = None,
    status: str | None = None,
    verifier: str | None = None,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticListResponse:
    """List diagnostic items with optional filters + approved coverage summary.

    Optional ``verifier=needs_review`` restricts to items with
    verifier_status IN (mismatch, ambiguous).
    """
    items, coverage = await list_items(
        session, market_code=market_code, topic=topic, status=status, verifier=verifier
    )
    return DiagnosticListResponse(
        items=[DiagnosticItemRead.model_validate(i) for i in items],
        coverage=coverage,
    )


# ---------------------------------------------------------------------------
# PATCH /admin/diagnostic-items/{id}
# ---------------------------------------------------------------------------


@router.patch("/diagnostic-items/{item_id}", response_model=DiagnosticItemRead)
async def edit_diagnostic_item(
    item_id: uuid.UUID,
    payload: DiagnosticItemPatch,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Edit a draft diagnostic item.  Returns 409 if not in draft status."""
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "draft":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot edit item with status '{item.status}'; only draft items may be edited",
        )
    item = await patch_item(
        session,
        item,
        fields_set=payload.model_fields_set,
        question=payload.question,
        choices=payload.choices,
        answer_index=payload.answer_index,
        explanation=payload.explanation,
        difficulty_tier=payload.difficulty_tier,
        concept_id=payload.concept_id,
    )
    await session.commit()
    return DiagnosticItemRead.model_validate(item)


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/{id}/approve
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/{item_id}/approve", response_model=DiagnosticItemRead)
async def approve_diagnostic_item(
    item_id: uuid.UUID,
    admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Transition a draft item to approved."""
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "draft":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot approve item with status '{item.status}'; must be draft",
        )
    item = await approve_item(session, item, admin_id=admin.id)
    await session.commit()
    return DiagnosticItemRead.model_validate(item)


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/{id}/reject
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/{item_id}/reject", response_model=DiagnosticItemRead)
async def reject_diagnostic_item(
    item_id: uuid.UUID,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Transition a draft item to retired (kept for audit)."""
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "draft":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot reject item with status '{item.status}'; must be draft",
        )
    item = await reject_item(session, item)
    await session.commit()
    return DiagnosticItemRead.model_validate(item)


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/{id}/retire
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/{item_id}/retire", response_model=DiagnosticItemRead)
async def retire_diagnostic_item(
    item_id: uuid.UUID,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Transition an approved item to retired."""
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "approved":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot retire item with status '{item.status}'; must be approved",
        )
    item = await retire_item(session, item)
    await session.commit()
    return DiagnosticItemRead.model_validate(item)


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/{id}/unpublish
# ---------------------------------------------------------------------------


@router.post("/diagnostic-items/{item_id}/unpublish", response_model=DiagnosticItemRead)
async def unpublish_diagnostic_item(
    item_id: uuid.UUID,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Transition an approved item back to draft so it can be corrected and re-approved."""
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "approved":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot unpublish item with status '{item.status}'; must be approved",
        )
    item = await unpublish_item(session, item)
    await session.commit()
    return DiagnosticItemRead.model_validate(item)


# ---------------------------------------------------------------------------
# POST /admin/diagnostic-items/{id}/clear-verifier-flag
# ---------------------------------------------------------------------------


@router.post(
    "/diagnostic-items/{item_id}/clear-verifier-flag",
    response_model=DiagnosticItemRead,
)
async def clear_diagnostic_verifier_flag(
    item_id: uuid.UUID,
    _admin: User = Depends(get_current_admin),
    session: AsyncSession = Depends(get_session),
) -> DiagnosticItemRead:
    """Dismiss an advisory verifier flag on an approved item judged correct.

    Clears the verifier_* fields in place — the item stays approved/published and
    its answer is untouched — so a false-positive flag drops out of needs_review.
    """
    item = await get_item(session, item_id)
    if item is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Item not found")
    if item.status != "approved":
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Cannot clear flag on item with status '{item.status}'; must be approved",
        )
    item = await clear_verifier_flag(session, item)
    await session.commit()
    return DiagnosticItemRead.model_validate(item)

File: frontend/src/api/adminDiagnostic.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiFetch } from './client';

// ── Types ──────────────────────────────────────────────────────────

export type VerifierStatus = 'agree' | 'mismatch' | 'ambiguous' | 'error';

export interface DiagnosticItem {
  id: string;
  market_code: string;
  topic: string;
  concept_id: string | null;
  difficulty_tier: 1 | 2 | 3;
  question: string;
  choices: string[];
  answer_index: number;
  explanation: string;
  status: 'draft' | 'approved' | 'retired';
  source: string;
  times_shown: number;
  times_correct: number;
  approved_by: string | null;
  approved_at: string | null;
  created_at: string;
  verifier_status: VerifierStatus | null;
  verifier_answer_index: number | null;
  verifier_note: string | null;
  verified_at: string | null;
}

export interface CoverageCell {
  topic: string;
  difficulty_tier: 1 | 2 | 3;
  approved_count: number;
}

export interface DiagnosticItemsResponse {
  items: DiagnosticItem[];
  coverage: CoverageCell[];
}

export interface GenerateParams {
  market_code: string;
  topic: string;
  difficulty_tier: 1 | 2 | 3;
  count: number;
}

export interface DiagnosticItemPatch {
  question?: string;
  choices?: string[];
  answer_index?: number;
  explanation?: string;
  difficulty_tier?: 1 | 2 | 3;
  concept_id?: string | null;
}

export interface DiagnosticFilters {
  market_code?: string;
  topic?: string;
  status?: string;
  verifier?: 'needs_review' | '';
}

export interface VerifyParams {
  market_code?: string;
  topic?: string;
  status?: string;
  limit?: number;
  only_unverified?: boolean;
  tier?: 1 | 2 | 3;
}

export interface VerifyFlaggedItem {
  id: string;
  topic: string;
  difficulty_tier: number;
  answer_index: number;
  verifier_answer_index: number | null;
  verifier_status: VerifierStatus;
  verifier_note: string | null;
}

export interface VerifyResult {
  verified: number;
  agree: number;
  mismatch: number;
  ambiguous: number;
  error: number;
  flagged: VerifyFlaggedItem[];
}

// ── Query keys ──────────────────────────────────────────────────────

const DIAGNOSTIC_KEY = ['admin', 'diagnostic-items'];

// ── Hooks ────────────────────────────────────────────────────────────

function useInvalidate() {
  const qc = useQueryClient();
  return () => qc.invalidateQueries({ queryKey: DIAGNOSTIC_KEY });
}

export function useDiagnosticItems(filters: DiagnosticFilters = {}) {
  const params = new URLSearchParams();
  if (filters.market_code) params.set('market_code', filters.market_code);
  if (filters.topic) params.set('topic', filters.topic);
  if (filters.status) params.set('status', filters.status);
  if (filters.verifier) params.set('verifier', filters.verifier);
  const qs = params.toString();
  return useQuery({
    queryKey: [...DIAGNOSTIC_KEY, filters],
    queryFn: () =>
      apiFetch<DiagnosticItemsResponse>(
        `/admin/diagnostic-items${qs ? `?${qs}` : ''}`,
      ) as Promise<DiagnosticItemsResponse>,
  });
}

export function useVerifyItems() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: (body: VerifyParams) =>
      apiFetch<VerifyResult>('/admin/diagnostic-items/verify', {
        method: 'POST',
        body: JSON.stringify(body),
      }) as Promise<VerifyResult>,
    onSuccess: invalidate,
  });
}

export function useGenerateItems() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: (body: GenerateParams) =>
      apiFetch<{ items: DiagnosticItem[] }>('/admin/diagnostic-items/generate', {
        method: 'POST',
        body: JSON.stringify(body),
      }),
    onSuccess: invalidate,
  });
}

export function usePatchItem() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: ({ id, body }: { id: string; body: DiagnosticItemPatch }) =>
      apiFetch<DiagnosticItem>(`/admin/diagnostic-items/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(body),
      }),
    onSuccess: invalidate,
  });
}

export function useApproveItem() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<DiagnosticItem>(`/admin/diagnostic-items/${id}/approve`, { method: 'POST' }),
    onSuccess: invalidate,
  });
}

export function useRejectItem() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<DiagnosticItem>(`/admin/diagnostic-items/${id}/reject`, { method: 'POST' }),
    onSuccess: invalidate,
  });
}

export function useRetireItem() {
  const invalidate = useInvalidate();
  return useMutation({
    mutationFn: (id: string) =>
      apiFetch<DiagnosticItem>(`/admin/diagnostic-items/${id}/retire`, { method: 'POST' }),
    onSuccess: invalidate,
  });
}
